### ISSUE TITLE
lookup public databases when creating new entities

### ISSUE THREAD

**Issue by moritzschepp** [coneda/kor#179]
Creating entities of a given kind should be possible via public databases like the GND or Wikidata: Per kind, there should be the possibility to add special fields which handle lookup in the database and suggest autocreation of the entry (including the database's identifier).

We should implement this after #155, otherwise we'll duplicate a lot of work.

We could access the GND data via http://blog.lobid.org/2018/06/07/lobid-gnd.html

**Comment by wuebbena** [comment]
Please have also  look at: https://twitter.com/ewg118/status/1005161946579161089

**Comment by moritzschepp** [comment]
I played around with wikidata's SPARQL endpoint and with the lobid api:

Consider this [query](https://query.wikidata.org/#SELECT%20%3Fwork%20%3Ftitle%20%3Ftype%20%3Fstatements%0AWHERE%0A%7B%0A%20%20%3Fwork%20wdt%3AP31%2Fwdt%3AP279%2a%20wd%3AQ838948%20.%0A%20%20%3Fwork%20wdt%3AP1476%20%3Ftitle%20.%0A%20%20OPTIONAL%20%7B%20%3Fwork%20wdt%3AP31%20%5Brdfs%3Alabel%20%24type%5D%20%7D.%0A%20%20OPTIONAL%20%7B%20%3Fwork%20wikibase%3Astatements%20%3Fstatements%7D%20.%0A%20%20FILTER%28REGEX%28STR%28%3Ftitle%29%2C%20%22%5E.%2amona%20lisa.%2a%24%22%2C%20%22i%22%29%29%20.%0A%20%20FILTER%28LANG%28%3Ftype%29%20%3D%20%22en%22%29%20.%0A%20%20FILTER%28LANG%28%3Ftitle%29%20%3D%20%22en%22%29%20.%0A%7D%0ALIMIT%20100%0AOFFSET%200). It finds the Q-ID, label, specific type and amount of statements for all instances of "work of art" filtered on the label by "mona lisa".

* without ordering, the search takes several seconds
* with ordering (on number of statements), time is increased to 30 seconds
* ordering by number of statements is not optimal and would position the painting "Mona Lisa" at position 6 or 7
* restricting to "paintings" would improve the run time but at the cost of a generic autocomplete

To a lesser extent, the same applies to the lobid API. Consider this [query](http://lobid.org/gnd/search?q=preferredName:leonardo%20AND%20type:Person&sort=preferredName.keyword%3Aasc&format=json). It lists all items with preferredName like "leonardo" of type "Person"

* its a lot faster than wikidata
* there is no suitable ranking that I could find (the above query results in 2740 results)

So bottom line: Neither of them would enable us to build a good user experience.

The wikidata [frontpage](https://wikidata.org) has a autocomplete which works exactly like we would want it and they use the mediawiki api behind the scenes. Will investigate if we can use that easily.

**Comment by moritzschepp** [comment]
Looking up data in wikidata has to be triggered for ALL related entities in addition to the current entity, because relationship data is one-way in wikidata. E.g. importing "Albrecht Dürer" (Q5580) and "Melencolia I" (Q1362177) yields different results depending on the order they were imported in.

**Comment by moritzschepp** [comment]
This is now fixed now.

TLDR: Turns out the above case is not only one to consider, e.g. using [Q571122](https://www.wikidata.org/wiki/Q571122) instead of Q1362177 is more difficult because it is not a "notable work" by Dürer and we don't find the work with the Dürer Wikidata item at all. I'm using the SPARQL endpoint to collect all items related to Q5580 but that in turn depends on a list of all previously imported Wikidata items. We have such a list (that's basically what the identifiers represent in KOR) but this might not perform well for larger installations.

**Comment by wuebbena** [comment]
So that's a new issue and we need a ticket for it?

**Comment by moritzschepp** [comment]
No, its just not a very elegant solution to the previous problem. Should we run into issues, we can always push it as a background process.

**Comment by wuebbena** [comment]
Ok, I  see. Here it is enough to stay alert.
